### PR TITLE
BUG #1944: Rename line to linetype for isoline, unified1d and vector.

### PR DESCRIPTION
**Pull request by danlipsa**
Add separate function to set type, color and width from a line.

### CODE REVIEW

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 Please review. This required more changes than I thought or wanted. Overall I think it is a much clearer API.

**Comment by sankhesh** [comment]
Do: test

**Comment by kwrobot01** [comment]
*Warning*: Requests to test ignored; user lacks testing privileges:  
@sankhesh: test   

**SERVER ERROR**: project uvcdat directory not available! @sankhesh

**Comment by kwrobot01** [comment]
*Warning*: Requests to test ignored; user lacks testing privileges:  
@sankhesh: test   

**SERVER ERROR**: project uvcdat directory not available! @sankhesh

**Comment by kwrobot01** [comment]
**SERVER ERROR**: project uvcdat directory not available! @sankhesh

**Comment by kwrobot01** [comment]
**SERVER ERROR**: project uvcdat directory not available! @sankhesh

**Comment by sankhesh** [comment]
Do: check

**Comment by danlipsa** [comment]
Do: check

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 ping! This also has been setting here for too long.

**Comment by danlipsa** [comment]
Do: test

**Comment by danlipsa** [comment]
@sankhesh @doutriaux1 This test
https://open.cdash.org/testDetails.php?test=462274077&build=4475133
uses
ACTIVATING ENV:2.6
but should be something like:
ACTIVATING ENV:2.6-11-g94c840d
Any idea why?

**Comment by doutriaux1** [comment]
That is odd indeed did you start with the 2.6 env ?

**Comment by danlipsa** [comment]
This is done by buildbot. Sankhesh says that he deletes both the source and the build directory so I am not sure where 2.6 is picked up from.

**Comment by danlipsa** [comment]
@doutriaux1 This is done by buildbot. Sankhesh says that he deletes both the source and the build directory so I am not sure where 2.6 is picked up from.

**Comment by danlipsa** [comment]
Do: test

**Comment by danlipsa** [comment]
Do: test

File: testing/vcs/test_vcs_set_line_attribute_from_line.py [hunk 1]
```diff
@@ -6,16 +6,22 @@
 l.type = "dash"
 
 v=vcs.createvector()
-v.line = "vcs_test_set_line"
+v.setLineAttributes("vcs_test_set_line")
```
Comment (aashish24) [inline testing/vcs/test_vcs_set_line_attribute_from_line.py]: @danlipsa what does this function do at this point in the code?

File: testing/vcs/test_vcs_set_line_attribute_from_line.py [hunk 2]
```diff
@@ -6,16 +6,22 @@
...
 
 assert(v.linecolor == 242)
 assert(v.linewidth == 5.6)
-assert(v.line == "dash")
+assert(v.linetype == "dash")
```
Comment (aashish24) [inline testing/vcs/test_vcs_set_line_attribute_from_line.py]: +1

File: Packages/vcs/vcs/vector.py [hunk 3]
```diff
@@ -507,14 +522,17 @@ def _setlinecolor(self, value):
...
+    def _setlinetype(self, value):
         if value is not None:
-            value = VCS_validation_functions.checkLineType(self, 'line', value)
-        self._line = value
-    line = property(_getline, _setline)
+            value = VCS_validation_functions.checkLineType(self, 'linetype', value)
+        self._linetype = value
+    linetype = property(_getlinetype, _setlinetype)
+
+    def setLineAttributes(self, line):
```
Comment (aashish24) [inline Packages/vcs/vcs/vector.py]: @danlipsa should we add docstring?

File: testing/vcs/test_vcs_set_line_attribute_from_line.py (hunk 1 above)
Comment (danlipsa) [inline testing/vcs/test_vcs_set_line_attribute_from_line.py]: Sets linetype, linecolor and linewidth from the line with the name 'vcs_test_set_line'.

File: Packages/vcs/vcs/vector.py (hunk 3 above)
Comment (danlipsa) [inline Packages/vcs/vcs/vector.py]: @aashish24 I missed that. If you search for 'def setLineAttributes' you'll find the docstring. I'll have to copy it in a few more places.

**Comment by danlipsa** [comment]
@aashish24 Done.

File: Packages/vcs/vcs/isoline.py [hunk 4]
```diff
@@ -619,28 +619,60 @@ def _setlinecolors(self, value):
...
-                self.linecolors = cvalue
-        except:
-            pass
-        try:
-            if self.linewidths is None or len(self.linewidths) < len(wvalue):
-                self.linewidths = wvalue
-        except:
-            pass
-
-    line = property(_getline, _setline)
```
Comment (doutriaux1) [inline Packages/vcs/vcs/isoline.py]: I don't think we should break backward compatibility, we should probably keep it and raise a deprecated warning.

File: Packages/vcs/vcs/unified1D.py [hunk 5]
```diff
@@ -46,7 +46,6 @@ def process_src(nm, code, typ):
                 "ymtics#1", "ymtics#2",
                 "xaxisconvert", "yaxisconvert",
                 "datawc_tunits",
-                "Tl",
```
Comment (doutriaux1) [inline Packages/vcs/vcs/unified1D.py]: you can't remove this, `scr` files are old files if the attribute is here that means it probably was part of the scr files.

File: Packages/vcs/vcs/isoline.py (hunk 4 above)
Comment (danlipsa) [inline Packages/vcs/vcs/isoline.py]: Makes sense. I will do that.

File: Packages/vcs/vcs/unified1D.py [hunk 6]
```diff
@@ -496,16 +510,21 @@ def _setlinecolor(self, value):
...
+        return self._linetype
 
-    def _setline(self, value):
+    def _setlinetype(self, value):
         if value is not None:
-            if vcs.queries.isline(value):
-                value = value.name
-            value = VCS_validation_functions.checkLineType(self, 'line', value)
-        self._line = value
-    line = property(_getline, _setline)
```
Comment (doutriaux1) [inline Packages/vcs/vcs/unified1D.py]: same here i think we should preserve backward compatibility for now and issue a deprecated warning

File: Packages/vcs/vcs/utils.py [hunk 7]
```diff
@@ -312,42 +312,6 @@ def _scriptrun(script, canvas=None):
                 vcs.process_src_element(s)
                 processing = False
     f.close()
-    # Ok now we need to double check the isolines
```
Comment (doutriaux1) [inline Packages/vcs/vcs/utils.py]: @danlipsa why is this removed?

File: Packages/vcs/vcs/vector.py [hunk 8]
```diff
@@ -59,9 +59,7 @@ def process_src(nm, code):
         sp = scode.split("=")
         nm = sp[0].strip()
         nm = nm.replace("#", "")
-        if nm == "Tl":
```
Comment (doutriaux1) [inline Packages/vcs/vcs/vector.py]: same here it will break reading in old `scr` files I think.

File: Packages/vcs/vcs/vector.py [hunk 9]
```diff
@@ -190,11 +205,11 @@ class Gv(object):
     vc.xyscale('linear', 'area_wt')     # Will set them both
 
     Specify the line style:
-     vc.line=0              # Same as vc.line='solid'
-     vc.line=1              # Same as vc.line='dash'
-     vc.line=2              # Same as vc.line='dot'
-     vc.line=3              # Same as vc.line='dash-dot'
-     vc.line=4              # Same as vc.line='long-dot'
+     vc.linetype=0              # Same as vc.linetype='solid'
```
Comment (doutriaux1) [inline Packages/vcs/vcs/vector.py]: should we simply set them to 'solid', etc.. it's easier to read no?

File: Packages/vcs/vcs/vector.py (hunk 3 above)
Comment (doutriaux1) [inline Packages/vcs/vcs/vector.py]: @danlipsa @aashish24 speaking of doc strings I asked our interns here to atually add them to the propertied objects as well. I will be a nice addition

File: testing/vcs/test_vcs_dump_json.json [hunk 10]
```diff
@@ -220,8 +224,8 @@
    "datawc_x2": 1e+20, 
    "datawc_y1": 1e+20, 
    "datawc_y2": 1e+20, 
-   "line": null, 
    "linecolor": null, 
+   "linetype": null,
```
Comment (doutriaux1) [inline testing/vcs/test_vcs_dump_json.json]: oh yes that is why we need to keep backward compatibilty for a while, all the json file have `line` in them. That will break them all.

File: testing/vcs/test_vcs_set_line_attribute_from_line.py [hunk 11]
```diff
@@ -6,16 +6,22 @@
...
-
+yx.setLineAttributes(l)
 
 assert(yx.linecolor == 242)
 assert(yx.linewidth == 5.6)
-assert(yx.line == "dash")
+assert(yx.linetype == "dash")
+
+iso = vcs.createisoline()
+iso.setLineAttributes([l, "solid", l])
```
Comment (doutriaux1) [inline testing/vcs/test_vcs_set_line_attribute_from_line.py]: maybe "solid" isn't the best choice here, it could llead people to think they can pass a line `type` instead of a line name. I would use something very different like: `thick` or `pink`

File: Packages/vcs/vcs/isoline.py (hunk 4 above)
Comment (danlipsa) [inline Packages/vcs/vcs/isoline.py]: @doutriaux1 @doutriaux1 Makes sense. The line interface will still have the bug @chaosphere2112 reported. We won't do any fixes there. Maybe remove it in the next version.

File: Packages/vcs/vcs/unified1D.py (hunk 5 above)
Comment (danlipsa) [inline Packages/vcs/vcs/unified1D.py]: @doutriaux1 Note that I deal with "Tl" later on, look at the full source code. That way I use the new interface not the deprecated one.

File: Packages/vcs/vcs/unified1D.py (hunk 6 above)
Comment (danlipsa) [inline Packages/vcs/vcs/unified1D.py]: @doutriaux1 Sounds good.

File: Packages/vcs/vcs/utils.py (hunk 7 above)
Comment (danlipsa) [inline Packages/vcs/vcs/utils.py]: @doutriaux1 I did not really understand that code, and something bad was happening there as far as I can remember. It made all tests pass. I will take a closer look.

File: Packages/vcs/vcs/vector.py (hunk 8 above)
Comment (danlipsa) [inline Packages/vcs/vcs/vector.py]: @doutriaux1 Same as before, I deal with 'Tl' later in the code using the new interface.

File: Packages/vcs/vcs/utils.py (hunk 7 above)
Comment (doutriaux1) [inline Packages/vcs/vcs/utils.py]: @danlipsa without taking a deep look I can tell you what I "remember" 
isoline graphic methods that had styles and colors used to be stored in the old `.scr` files as separte `line` objects.
I think this bit of code as something to do with this.

File: Packages/vcs/vcs/vector.py (hunk 9 above)
Comment (danlipsa) [inline Packages/vcs/vcs/vector.py]: @doutriaux1 This is a comment in the sources, the user should use strings instead of numbers indeed.

File: testing/vcs/test_vcs_set_line_attribute_from_line.py (hunk 11 above)
Comment (danlipsa) [inline testing/vcs/test_vcs_set_line_attribute_from_line.py]: I'll just add a comment.

File: Packages/vcs/vcs/utils.py [hunk 12]
```diff
@@ -312,42 +312,6 @@ def _scriptrun(script, canvas=None):
...
-    gd = vcs.elements["isoline"]["default"]
-    for g in vcs.elements["isoline"].values():
-        if g.name == "default":
-            continue
-        for att in ["line", "textcolors", "text"]:
-            try:
-                setattr(g, att, getattr(g, att))
```
Comment (danlipsa) [inline Packages/vcs/vcs/utils.py]: @doutriaux1 What does this code do?

File: Packages/vcs/vcs/utils.py [hunk 13]
```diff
@@ -312,42 +312,6 @@ def _scriptrun(script, canvas=None):
...
-                lst = []
-                if att == "line":
-                    for e in g.line:
-                        if e in vcs.elements["line"]:
-                            lst.append(vcs.elements["line"][e])
-                        else:
-                            lst.append(e)
-                elif att == "text":
-                    for e in g.line:
-                        if e in vcs.elements["textorientation"]:
```
Comment (danlipsa) [inline Packages/vcs/vcs/utils.py]: @doutriaux1 Why is e from g.line is search in vcs.elements['textorientation']?

File: Packages/vcs/vcs/utils.py [hunk 14]
```diff
@@ -312,42 +312,6 @@ def _scriptrun(script, canvas=None):
...
-                            lst.append(vcs.elements["line"][e])
-                        elif e in vcs.elements["textcombined"]:
-                            lst.append(vcs.elements["line"][e])
-                        else:
-                            lst.append(e)
-                elif att == "textcolors":
-                    for e in g.line:
-                        if e in vcs.elements["texttable"]:
```
Comment (danlipsa) [inline Packages/vcs/vcs/utils.py]: @doutriaux1 Same question here: what is the link between e from g.line (which is a linestyle) and vcs.elements['texttable']?

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 I still don't understand the code I removed from utils.py - it does not affect the tests. Maybe you guys can help me here. I added back the line interface which prints a deprecated message and calls on the new linetype/setLineAttributes interface. Please review

**Comment by danlipsa** [comment]
These are the failures on linux:
29 tests failed out of 611:

	 13 - testWrite (Failed)
	 62 - test_vcs_read_old_scr_2 (Failed)
	 73 - test_vcs_verify_boxfill_basics (Failed)
	121 - test_vcs_boxfill_decreasing_latitude (Failed)
	122 - test_vcs_meshfill_no_wrapping (Failed)
	413 - test_vcs_isofill_level0 (Failed)
	414 - test_vcs_isofill_level1 (Failed)
	415 - test_vcs_isofill_level2 (Failed)
	469 - test_vcs_matplotlib_colormap (Failed)
	549 - dv3d_vector_test (Failed)
	550 - dv3d_slider_test (Failed)
	551 - Hovmoller_volume_test (Failed)
	552 - dv3d_volume_test (Failed)
	553 - dv3d_constituents_test (Failed)
	554 - dv3d_surface_test (Failed)
	555 - dv3d_hovmoller_test (Failed)
	580 - cdms2_test_createcopy_lose_dtype (Failed)
	581 - CDMS_Test_01 (Failed)
	582 - CDMS_Test_02 (Failed)
	583 - CDMS_Test_03 (Failed)
	584 - CDMS_Test_04 (Failed)
	590 - CDMS_Test_10 (Failed)
	598 - CDMS_Test_18 (Failed)
	599 - CDMS_Test_20 (Failed)
	601 - CDMS_Test_multiple_formats (OTHER_FAULT)
	602 - CDMS_Test_write_compressed (Failed)
	606 - CDMS_Test_pack (Failed)
	609 - CDMS_Test_write_extended_dimension (Failed)
	610 - CDMS_Test_del_attributes (Failed)

File: Packages/vcs/vcs/utils.py (hunk 13 above)
Comment (doutriaux1) [inline Packages/vcs/vcs/utils.py]: because textorientation objects contain the relevant information to this attribute so it is possible that the use rsent us a list of textorientation to look for. We could probably drop this and the `texttable` bellow and simply look in `textcombined` it would be easier to understand and maintain.

File: Packages/vcs/vcs/utils.py (hunk 14 above)
Comment (doutriaux1) [inline Packages/vcs/vcs/utils.py]: i think you're right it should be `g.text` and `g.textcolor` looks like some copy/pasting that went bad.

**Comment by doutriaux1** [comment]
@danlipsa I will look into this, but it maybe easier to look at this together when you're here next week, I'm not sure why you removed it at all.

**Comment by danlipsa** [comment]
Do: test

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 This is ready! I added back the removed code and the line interface.

**Comment by danlipsa** [comment]
This are the test failures on Linux:
96% tests passed, 27 tests failed out of 611

Total Test time (real) = 407.44 sec

The following tests FAILED:
	 13 - testWrite (Failed)
	 73 - test_vcs_verify_boxfill_basics (Failed)
	121 - test_vcs_boxfill_decreasing_latitude (Failed)
	122 - test_vcs_meshfill_no_wrapping (Failed)
	413 - test_vcs_isofill_level0 (Failed)
	414 - test_vcs_isofill_level1 (Failed)
	415 - test_vcs_isofill_level2 (Failed)
	469 - test_vcs_matplotlib_colormap (Failed)
	549 - dv3d_vector_test (Failed)
	550 - dv3d_slider_test (Failed)
	551 - Hovmoller_volume_test (Failed)
	552 - dv3d_volume_test (Failed)
	553 - dv3d_constituents_test (Failed)
	554 - dv3d_surface_test (Failed)
	555 - dv3d_hovmoller_test (Failed)
	580 - cdms2_test_createcopy_lose_dtype (Failed)
	581 - CDMS_Test_01 (Failed)
	582 - CDMS_Test_02 (Failed)
	583 - CDMS_Test_03 (Failed)
	584 - CDMS_Test_04 (Failed)
	590 - CDMS_Test_10 (Failed)
	598 - CDMS_Test_18 (Failed)
	599 - CDMS_Test_20 (Failed)
	602 - CDMS_Test_write_compressed (Failed)
	606 - CDMS_Test_pack (Failed)
	609 - CDMS_Test_write_extended_dimension (Failed)
	610 - CDMS_Test_del_attributes (Failed)

**Comment by doutriaux1** [comment]
@danlipsa no worries about the failed test, it's due to hdf5 not matching anymore by default under conda. But once I merge the legend PR you will likely need to re-update your baselines.

**Comment by danlipsa** [comment]
@doutriaux1 Indeed the failed vcs tests are because of hdf5 problem.

**Comment by danlipsa** [comment]
@doutriaux1 @aashish24 Lets get this merged as well, unless you have additional comments/concerns.

**Comment by doutriaux1** [comment]
@danlipsa I would like to finish/merge fix_travis first if it doesn't take too long.

**Comment by danlipsa** [comment]
@doutriaux1 Do you worry about conflicts/broken tests?

**Comment by doutriaux1** [comment]
kind of the opposite, since it shouldn't break anything I would like to see a PR pass on travis ;)

**Comment by danlipsa** [comment]
@doutriaux1 I guess my point was that even if you merge this your PR on travis can go on as before. You should not have any conflicts or any new failing tests. But I can wait if you want to take a closer look at this and you want to finish the travis work first without interruption.

**Comment by danlipsa** [comment]
Do: check

**Comment by danlipsa** [comment]
Do: test

**Comment by danlipsa** [comment]
Do: test

**Comment by sankhesh** [comment]
Do: test

**Comment by aashish24** [comment]
@sankhesh I see 4 contents check but no build / test check after the last test command from @danlipsa. What's wrong?

**Comment by sankhesh** [comment]
@aashish24 I was puzzled by it yesterday too. I looked into it. This pull request is older than the changes we made to the bot and I am assuming buildbot is getting confused with the different commands on it. I'll talk to Ben on Monday to see what's wrong exactly.

**Comment by sankhesh** [comment]
Do: test

**Comment by sankhesh** [comment]
Do: test

**Comment by danlipsa** [comment]
@sankhesh It seems that some of the tests still do not activate the right environment:
https://open.cdash.org/testDetails.php?test=462274077&build=4495107

**Comment by sankhesh** [comment]
@danlipsa All the tests on default build are activating the right environment. The tests on the lean build just activate 2.6. I think that is a UV-CDAT issue that needs to be fixed in the code.

**Comment by danlipsa** [comment]
@sankhesh How do you see the default build? I only see the lean build - I click on cdash

**Comment by sankhesh** [comment]
Here is the default build: https://open.cdash.org/buildSummary.php?buildid=4495026

**Comment by danlipsa** [comment]
@sankhesh The test failures look like @doutriaux1 colorbar baselines. I will run the test on my machine to see if they pass.

**Comment by danlipsa** [comment]
@sankhesh On my machine most tests pass - I have 26 failures - most vcs once are from hdf5.
Here are @doutriaux1 baselines.
https://github.com/UV-CDAT/uvcdat-testdata/pull/149/files

In the link you provided, it looks like @doutriaux1 changes are not in, possible because the code us the wrong conda env? The SHA is correct. The baseline is right but the generated image is wrong. Can you look in the LastTest.log to see the conda env for the tests?

**Comment by doutriaux1** [comment]
Do: test
